### ISSUE TITLE
Cargo dependency versions are not locked

### ISSUE THREAD

**Issue by shaleh** [Wilfred/remacs#508]
I have been wondering why I keep seeing `Cargo.lock` as an edited file. Today I looked into it.

```
foo = "0.1.1"
```

Does not mean "get version 0.1.1" it means "get version >= 0.1.1". If we actually mean version "0.1.1" we have to use an `=`.

```
foo = "=0.1.1"
```

Is our intent to lock the version?

**Comment by shaleh** [comment]
lazy_static was one that stood out:

    -lazy_static = "0.2.2"
    +lazy_static = "=0.2.11"

**Comment by Wilfred** [comment]
We do have Cargo.lock checked in, so the build should be reproducible. What command is updating Cargo.lok?

**Comment by shaleh** [comment]
Forgive me for not explaining it well.

If the version in Cargo.toml does not have an equal sign then when a build is run cargo will download the latest version greater than or equal to the one listed.

**Comment by shaleh** [comment]
Look at the versions in the checked in Cargo.lock versus what is in the Cargo.toml.

**Comment by shaleh** [comment]
Today I was able to run this

```
#!/bin/sh

set +x

git clone https://github.com/Wilfred/remacs
cd remacs
./autogen.sh
./configure --enable-rust-debug
make -j4
git status
```

a week or so ago it showed changes to Cargo.lock.

**Comment by jaccarmac** [comment]
According to the docs for Cargo, the lock file should only get updated when the manifest is changed. Otherwise you're seeing a bug.

**Comment by shaleh** [comment]
Right. So make clean in a tree you have had for a bit. Then build again. This time there is 0.1.35 of crate X where you have 0.1.32. It will be downloaded and built and Cargo.lock will be updated. This is mostly expected behavior.

My point is we are loose with our versions which leads to Cargo.lock updates that we did not expect.

**Comment by jaccarmac** [comment]
Interesting. My tree isn't out of date so I can't test easily, but I can't see any commits in the past little while where a lock file changes without a change to the non-locked manifest. Something I did notice, however, is that the top-level `Cargo.lock` seems to get updated when child `Cargo.toml` files change.
